### PR TITLE
Use TestUtils.waitForAndAssertCondition instead of DisplayHelper

### PR DESCRIPTION
**Pull request by sebthom**
This PR increases the comprehensibility of test cases by using `TestUtils.waitForAndAssertCondition` instead of `new DisplayHelper() { ... }.waitForCondition` + `Assert.assertTrue`  which needs a lot of boilerplate.

While doing the changes I already identified a test case which is not really working `LanguageServiceAccessorTest#testReuseMultirootFolderLSAfterInitialProjectGotDeleted`. I added TODO/FIXME tags accordingly since I don't know how to fix it.

### CODE REVIEW

File: org.eclipse.lsp4e.test/src/org/eclipse/lsp4e/test/TestUtils.java [hunk 1]
```diff
@@ -233,17 +237,66 @@ public static IEditorReference[] getEditors() {
...
+		assertTrue("Condition not met within expected time.",
+				waitForCondition(timeout_ms, display, condition));
+	}
+
+	public static void waitForAndAssertCondition(String errorMessage, int timeout_ms, Condition condition) {
+		assertTrue(errorMessage, waitForCondition(timeout_ms, condition));
+	}
+
+	public static void waitForAndAssertCondition(String errorMessage, int timeout_ms, Display display, Condition condition) {
+		var ex = new Exception[1];
```
Comment (rubenporras) [inline org.eclipse.lsp4e.test/src/org/eclipse/lsp4e/test/TestUtils.java:255]: why do we use an array with one position instead of just a simple variable?

File: org.eclipse.lsp4e.test/src/org/eclipse/lsp4e/test/TestUtils.java [hunk 2]
```diff
@@ -233,17 +237,66 @@ public static IEditorReference[] getEditors() {
...
+		var isConditionMet = new DisplayHelper() {
```
Comment (rubenporras) [inline org.eclipse.lsp4e.test/src/org/eclipse/lsp4e/test/TestUtils.java:256]: this is just my taste, but I would use in general `val` for things which can be final.

File: org.eclipse.lsp4e.test/src/org/eclipse/lsp4e/test/TestUtils.java [hunk 3]
```diff
@@ -233,17 +237,66 @@ public static IEditorReference[] getEditors() {
...
+	}
+
+	public static boolean waitForCondition(int timeout_ms, Condition condition) {
 		return waitForCondition(timeout_ms, PlatformUI.getWorkbench().getDisplay(), condition);
 	}
 
-	public static boolean waitForCondition(int timeout_ms, Display display, BooleanSupplier condition) {
-		return new DisplayHelper() {
+	public static boolean waitForCondition(int timeout_ms, Display display, Condition condition) {
+		var ex = new Exception[1];
```
Comment (rubenporras) [inline org.eclipse.lsp4e.test/src/org/eclipse/lsp4e/test/TestUtils.java:281]: same remark as above, why an array?

File: org.eclipse.lsp4e.test/src/org/eclipse/lsp4e/test/LanguageServiceAccessorTest.java [hunk 4]
```diff
@@ -185,7 +187,7 @@ public void testGetOnlyRunningLanguageServers() throws Exception {
 		editor1 = openEditor(testFile1);
 		assertNotEmpty(getInitializedLanguageServers(testFile1, MATCH_ALL));
 
-		waitForCondition(5_000, () -> getActiveLanguageServers(MATCH_ALL).size() == 1);
+		waitForCondition(5_000, () -> getActiveLanguageServers(MATCH_ALL).size() > 0);
```
Comment (rubenporras) [inline org.eclipse.lsp4e.test/src/org/eclipse/lsp4e/test/LanguageServiceAccessorTest.java:190]: why this change?

File: org.eclipse.lsp4e.test/src/org/eclipse/lsp4e/test/TestUtils.java (hunk 2 above)
Comment (sebthom) [reply org.eclipse.lsp4e.test/src/org/eclipse/lsp4e/test/TestUtils.java:256]: there is no val, this isn't scala/kotlin :-)

File: org.eclipse.lsp4e.tests.mock/src/org/eclipse/lsp4e/tests/mock/MockLanguageServerMultiRootFolders.java [hunk 5]
```diff
@@ -64,7 +64,7 @@ public final class MockLanguageServerMultiRootFolders implements LanguageServer
 	private MockWorkspaceService workspaceService = new MockWorkspaceService(this::buildMaybeDelayedFuture);
 	private InitializeResult initializeResult = new InitializeResult();
 	private long delay = 0;
-	private boolean started;
+	private volatile boolean started;
```
Comment (rubenporras) [inline org.eclipse.lsp4e.tests.mock/src/org/eclipse/lsp4e/tests/mock/MockLanguageServerMultiRootFolders.java:67]: why this change?

File: org.eclipse.lsp4e.test/src/org/eclipse/lsp4e/test/TestUtils.java (hunk 2 above)
Comment (rubenporras) [reply org.eclipse.lsp4e.test/src/org/eclipse/lsp4e/test/TestUtils.java:256]: alright :-)

File: org.eclipse.lsp4e.test/src/org/eclipse/lsp4e/test/TestUtils.java (hunk 1 above)
Comment (sebthom) [reply org.eclipse.lsp4e.test/src/org/eclipse/lsp4e/test/TestUtils.java:255]: this stems from the fact that inner anonymous/lambdas instances can only access final local fields but we need to write to a field.

File: org.eclipse.lsp4e.tests.mock/src/org/eclipse/lsp4e/tests/mock/MockLanguageServerMultiRootFolders.java (hunk 5 above)
Comment (sebthom) [reply org.eclipse.lsp4e.tests.mock/src/org/eclipse/lsp4e/tests/mock/MockLanguageServerMultiRootFolders.java:67]: because we have code that runs in other threads which queries the isStarted() method. without the volatile it is not guaranteed that these threads see the current value.

**Review by rubenporras — APPROVED** [review]

**Comment by rubenporras** [comment]
@sebthom , shall I merge?

**Comment by sebthom** [comment]
> @sebthom , shall I merge?

Sure, go for it. Thanks.
